### PR TITLE
sketch at issue 317.

### PR DESCRIPTION
**Pull request by bradtofel**
I was mostly focusing on not breaking existing users.

The other change I thought of going with would be adding a method to QueryParam:

public PaginationValue getPaginationValue(RestrictedPaginationKeys key)

But what I have here seems more in line with the other QueryParam.get*() methods.

The biggest change here I see is that the NumberFormatException would be thrown in "user" Repository code, instead of higher up in the query parsing code.

### CODE REVIEW

**Comment by ieugen** [comment]
Hi @bradtofel , 

Thanks for the PR. I am in favor of adding more utility classes to help simplify working with query parameters. 

Regarding your PR, the helpers are great, however I would not add a wrapper since it creates more objects per request. That will not scale very well if you have a lot of users. 
I would recommend adding methods someplace else. Instead of wrapping it maybe adding them to the QueryParams class as helpers to request parameters as specific types. This is just an idea. 

Don't worry too much about breaking. We are planning to release 3.0.0. soon (or maybe 2.5 ??) It will not be that big of a change API wise for end users.  I've re-written a lot of the internals and the integrations will need to be updated.

**Comment by bradtofel** [comment]
Thanks for the suggestions @ieugen and also for the info about the 3.0.0 release. I want to make sure I understand your points.

Are you suggesting we remove the PaginationValue wrapper class, to keep object creation down, or do you mean some other wrapper class?

Since there are only likely to be 2 of these per request, my temptation would be to lean towards making the API cleaner, but I'm happy to pull it out if you think that's the way to go.

**Comment by ieugen** [comment]
I'm suggesting dropping the wrapper and storing the raw string. 

Instead of:
````
result.getPaginationValues()
                .get(RestrictedPaginationKeys.limit).asInteger())
````
you could change the API and add public API methods on the QueryParams to retrieve the pagination values for each supported type. 
````
result.getPaginationAsInt(RestrictedPaginationKeys.limit)
or a  generic form
result.getPaginationAs(RestrictedPaginationKeys.limit, Integer.class)
````
2 objects don't seem much, but they tend to add on the long run. 

Performance wise, is not that much for most clients, but it really depends on how many requests you get. 

Anyway, good quality software should work great regardless of the number of requests.

**Comment by bradtofel** [comment]
Got it, thanks. I'll amend the PR.

File: src/main/java/io/katharsis/queryParams/QueryParams.java [hunk 1]
```diff
@@ -226,21 +227,47 @@ void setGrouping(Map<String, Set<String>> grouping) {
...
-    public Map<RestrictedPaginationKeys, Integer> getPagination() {
+    public Map<RestrictedPaginationKeys, String> getPagination() {
         return pagination;
     }
 
+    /**
+     * @return the request parameter value for key converted to an Integer, or null if the parameter was not provided
+     * @throws NumberFormatException if the parameter cannot be converted to an Integer
+     */
+    public Integer getPaginationAsInt(RestrictedPaginationKeys key) {
```
Comment (bradtofel) [inline src/main/java/io/katharsis/queryParams/QueryParams.java]: @ieugen , I was about to return 'int' and 'long' here, but returning null when the param is missing seemed better than some "missing" value.. Doing this seems like we've lost any object creation advantages, though. Thoughts?

Comment (ieugen) [inline src/main/java/io/katharsis/queryParams/QueryParams.java]: One option is Optional. That way you force the user to make a decision. It's still a wrapper but it adds meaning to the code.

Comment (bradtofel) [inline src/main/java/io/katharsis/queryParams/QueryParams.java]: So I'm losing the thread here. The issue with the original PR was that it constructed an extra 2 objects per request, to wrap the String values. To me, that code looked much more congruent interface-wise to the rest of the accessors on QueryParams. I understand the uses of Optional, but is the goal here still to reduce objection-creation? I'm happy to add the Optionals, but this just makes more objects.

**Pull request merged**